Project 1 - Bach Chorales
---
By: Samantha M. Lord

User/Avatar ID : Salmondita007


Here I will be downloading and unzipping the dataset that will be assessed. 
--
I will be using Linux commands that we learned in previous labs

In [1]:
!curl -o bach.zip https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip

!unzip bach.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 41761  100 41761    0     0   254k      0 --:--:-- --:--:-- --:--:--  254k
Archive:  bach.zip
  inflating: bach.csv                


Importing the dataset that will be assessed using Pandas and then displaying it

In [13]:
import pandas as pd
bach = pd.read_csv('bach.csv')
bach

choral_ID  event_number    C  C#    D  D#    E    F  F#    G  G#    A  \
0     000106b_             1  YES  NO   NO  NO   NO  YES  NO   NO  NO  YES   
1     000106b_             2  YES  NO   NO  NO  YES   NO  NO  YES  NO   NO   
2     000106b_             3  YES  NO   NO  NO  YES   NO  NO  YES  NO   NO   
3     000106b_             4  YES  NO   NO  NO   NO  YES  NO   NO  NO  YES   
4     000106b_             5  YES  NO   NO  NO   NO  YES  NO   NO  NO  YES   
...        ...           ...  ...  ..  ...  ..  ...  ...  ..  ...  ..  ...   
5660  015505b_           105   NO  NO  YES  NO   NO   NO  NO  YES  NO   NO   
5661  015505b_           106   NO  NO  YES  NO   NO   NO  NO  YES  NO  YES   
5662  015505b_           107  YES  NO   NO  NO  YES   NO  NO  YES  NO   NO   
5663  015505b_           108  YES  NO   NO  NO  YES   NO  NO  YES  NO   NO   
5664  015505b_           109   NO  NO   NO  NO   NO  YES  NO   NO  NO  YES   

       A#   B bass  meter chord_label  
0      NO  NO    F      3         F_M  
1      NO  NO    E      5         C_M  
2      NO  NO    E      2         C_M  
3      NO  NO    F      3         F_M  
4      NO  NO    F      2         F_M  
...   ...  ..  ...    ...         ...  
5660  YES  NO    G      4         G_m  
5661   NO  NO    G      3         G_m  
5662   NO  NO    C      5         C_M  
5663  YES  NO    C      3         C_M  
5664   NO  NO    F      4         F_M  

[5665 rows x 17 columns]

Next I will be converting the features into integers. Then Splitting the datasets, training and testing them

In [17]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

columns = ['C','C#','D','D#','E','F','F#','G','G#','A','A#','B','bass','chord_label'] # Make the catagorial features read as numbers. 
le = LabelEncoder()
bach[columns] = bach[columns].apply(lambda col: le.fit_transform(col))

bach.set_index('choral_ID', inplace=True)
bFeatures = bach.drop('chord_label', axis=1)
bLabels = bach['chord_label']
bach_train_features, bach_test_features, bach_train_labels, bach_test_labels = train_test_split(bFeatures, bLabels, test_size = 0.2, random_state=42)


Now I will be using the Forest Classifier

In [19]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100, random_state=42)
rfc.fit(bach_train_features, bach_train_labels)

RandomForestClassifier(random_state=42)

Here I will be making predictions using the KNNClassifier

In [20]:
from sklearn.neighbors import KNeighborsClassifier

knnOne = KNeighborsClassifier(n_neighbors=4, p= 1, weights='distance')
knnOne.fit(bach_train_features, bach_train_labels)

predictions = knnOne.predict(bach_test_features)
predictions

array([64, 94, 52, ..., 35, 81, 78])

Lastly, I will be obtaining the accuracy 

In [22]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(bach_test_labels, predictions)
print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 70.26%


The accuracy is 70.26%. Which is a good reading of accuracy. We however, want to get it as close to 1 as possible to give the highest form of accuracy. This can happen because of the data that is given. The data isnt always apt for testing and training, it has to be modified to give the desired results. 